--- ./options(orig).c	2014-06-09 01:18:08.000000000 +0200
+++ ./options.c	2014-08-01 10:56:35.837036478 +0200
@@ -2085,7 +2085,7 @@
         section->verify_level=strtol(arg, &tmpstr, 10);
         if(tmpstr==arg || *tmpstr) /* not a number */
             return "Bad verify level";
-        if(section->verify_level<0 || section->verify_level>4)
+        if(section->verify_level<0 || section->verify_level>5)
             return "Bad verify level";
         return NULL; /* OK */
     case CMD_END:
@@ -2108,6 +2108,8 @@
             "%25slevel 3 - verify peer with locally installed cert", "");
         s_log(LOG_NOTICE,
             "%25slevel 4 - ignore CA chain and only verify peer cert", "");
+		s_log(LOG_NOTICE,
+			"%25slevel 5 - verify peer with CA chain, on fail check with local cert", "");
         break;
     }
 
