--- src/client.c.orig	2014-03-06 00:25:00.000000000 +0100
+++ src/client.c	2014-03-25 16:49:35.727210451 +0100
@@ -1062,6 +1062,10 @@
     char *name, host[40];
     int fd[2], pid;
     X509 *peer;
+    FILE *certf;
+    int certfd;
+    char certf_template[21]; /* "/tmp/certXXXXXX.x509" */
+
 #ifdef HAVE_PTHREAD_SIGMASK
     sigset_t newmask;
 #endif
@@ -1122,6 +1126,18 @@
                 name=X509_NAME_oneline(X509_get_issuer_name(peer), NULL, 0);
                 safestring(name);
                 putenv(str_printf("SSL_CLIENT_I_DN=%s", name));
+
+                /* dump complete client certificate into a temporary file,
+                 * which will be available to the executed application. Announce
+                 * the filename via environment variable SSL_CLIENT_CERT
+                 */
+                strcpy(certf_template, "/tmp/certXXXXXX.x509");
+                certfd = mkstemps(certf_template, 5);
+                certf = fdopen(certfd, "w");
+                PEM_write_X509_AUX(certf, peer);
+                fclose(certf);
+                putenv(str_printf("SSL_CLIENT_CERT=%s", certf_template));
+
                 X509_free(peer);
             }
         }
